### PR TITLE
Refactor git hook orchestration

### LINKED ISSUE #395: Ignore repos list
There should be a config option for which repos to ignore events from. `apertium-all`, `apertium-trunk`, etc.

### PR DESCRIPTION
**Pull request by Androbin**
Fixes #395

### CODE REVIEW

**Comment by Androbin** [comment]
@ftyers If you can express our needs in this form, I can implement that:
( repo_name, branch_name, is_fork, event_type ) -> list_of_channels

**Comment by sushain97** [comment]
Unfortunately, that isn't great. It would take a 500+ long config. If the repo name could be a regex, that would solve it.

**Comment by Androbin** [comment]
> Unfortunately, that isn't great. It would take a 500+ line long config.

However the exact solution, this is the information we need to be able to extract / construct.

> If the repo name could be a regex, that would solve it.

Yeah, but wouldn't a list suffice?

**Comment by Androbin** [comment]
Example configuration (for last commit):
```
git_events = {'apertium/phenny': ['issue', 'push'], 'apertium/apertium_apy': 'issue'}
git_channels = {'*': '#apertium', 'issue': '#apertium_issues', 'push': '#apertium_pushes'}
```

**Comment by Androbin** [comment]
@sushain97 Would this be a sensible representation of our needs?

**Comment by sushain97** [comment]
```
git_events = {
  'apertium/phenny': ['issue', 'push'],
  'apertium/apertium_apy': ['issue']
}
```

Can we have wildcards here? And override wildcards as well? We want issue/push from apertium/* but nothing from apertium/apertium-all, apertium/apertium-incubator, etc.

**Comment by Androbin** [comment]
We can use `fnmatch` for Unix shell-style wildcards.
And it will iterate over the entries, so order will be important.

**Comment by Androbin** [comment]
Can now be as crazy as:
```
git_events = {
  '*': ['push'],
  'apertium/*': ['issue', 'push'],
  'apertium/apertium-*': ['issue'],
  'apertium/apertium-all': [],
  'apertium/apertium-incubator': [],
}
```

**Comment by sushain97** [comment]
That seems like what we want...

**Comment by ftyers** [comment]
@Androbin that looks nice.

File: modules/git.py [hunk 1]
```diff
@@ -12,6 +12,7 @@
 import time
 import atexit
 import signal
+from fnmatch import fnmatch
```
Comment (sushain97) [inline modules/git.py]: Do you mind sorting and splitting into external/internal?

File: modules/git.py [hunk 2]
```diff
@@ -129,7 +130,11 @@ def do_POST(self):
             return
 
         try:
-            self.do_POST_unsafe(data)
+            if self.do_POST_unsafe(data):
+                # send OK code
+                self.send_response(200)
+                self.send_header("Content-type", "text/html")
```
Comment (sushain97) [inline modules/git.py]: I don't think these headers are necessary?

File: modules/git.py [hunk 3]
```diff
@@ -149,14 +154,18 @@ def do_POST(self):
 
     def do_POST_unsafe(self, data):
         # will contain both commit reports and error reports
-        msgs_by_channel = {}
-        msgs_default_channels = []
+        messages = []
+
+        try:
+            default_channels = self.phenny.config.git_channels['*']
```
Comment (sushain97) [inline modules/git.py]: the `try`-`except` is pretty verbose. prefer:

`self.phenny.config.git_channels.get('*', self.phenny.config.channels)

File: modules/git.py [hunk 4]
```diff
@@ -167,6 +176,26 @@ def do_POST_unsafe(self, data):
             elif 'organization' in data:
                 repo = data['organization']['login'] + ' (org)'
 
+            if self.phenny.config.git_events:
+                full_name = data['repository']['full_name']
+
+                for key, value in self.phenny.config.git_events.items():
+                    if fnmatch(full_name, key):
+                        event_types = value
+                        break
```
Comment (sushain97) [inline modules/git.py]: Won't this always result in us picking the least specific instead of the most specific?

File: modules/git.py [hunk 5]
```diff
@@ -167,6 +176,26 @@ def do_POST_unsafe(self, data):
...
+                else:
+                    event_types = []
+            else:
+                event_types = None
+
+            if event_types and (event not in event_types):
+                return True
+
+            try:
+                channels = self.phenny.config.git_channels[event]
```
Comment (sushain97) [inline modules/git.py]: `channels = self.phenny.config.git_channels.get(event, channels)`

File: modules/git.py [hunk 6]
```diff
@@ -326,58 +344,42 @@ def do_POST_unsafe(self, data):
...
-                if channel in msgs_by_channel:
-                    msgs_by_channel[channel].append(message)
-                else:
-                    msgs_by_channel[channel] = [message]
+        # post all messages to all channels
 
-        for channel in msgs_by_channel.keys():
-            more.add_messages(self.phenny, channel, msgs_by_channel[channel])
+        for channel in channels:
+            more.add_messages(self.phenny, channel, messages)
```
Comment (sushain97) [inline modules/git.py]: Won't this end up sending the same set of messages to all the channels?

Comment (Androbin) [reply modules/git.py]: The hook runs once per event. One repository, one event type.

File: modules/git.py (hunk 4 above)
Comment (Androbin) [reply modules/git.py]: Touchê

File: modules/git.py (hunk 2 above)
Comment (Androbin) [reply modules/git.py]: Touchê

File: modules/git.py (hunk 6 above)
Comment (sushain97) [reply modules/git.py]: Ah, yes. OK.

File: modules/git.py (hunk 1 above)
Comment (Androbin) [reply modules/git.py]: Done

File: modules/git.py (hunk 3 above)
Comment (Androbin) [reply modules/git.py]: Done

File: modules/git.py (hunk 5 above)
Comment (Androbin) [reply modules/git.py]: Done

**Pull request merged**